File: packages/go/qorechain/signbytes/signbytes.go
// Package signbytes builds the exact byte strings a QoreChain post-quantum key
// signs, in both forms the chain has used, and decides which form a given
// network verifies.
//
// Chain v3.1.98 introduced a second ("v2") form of three sign-bytes:
//
//   - hybrid tx (the ML-DSA-87 half of a hybrid classical + PQC transaction)
//   - PQC key migration (MsgMigratePQCKey, signed by both the old and new key)
//   - bridge attestation (MsgBridgeAttestation, signed by bridge validators)
//
// v2 adds a domain tag and binds the chain id, so a signature produced in one
// context (or for one network) never verifies in another. Each network accepts
// exactly ONE form at any height, with no overlap window:
//
//   - a network that existed before v3.1.98 ("qorechain-vladi" mainnet,
//     "qorechain-diana" testnet) verifies v1 until the v3.1.98 upgrade plan is
//     applied on it, and v2 from then on;
//   - any other chain id starts on a binary that verifies v2 from block one.
//
// The testnet applied v3.1.98 at height 5,746,000; mainnet stays on v1 until its
// own governance upgrade. Clients therefore must not hardcode a form: use
// VersionFor, or a Resolver that asks the node whether the plan is applied.
//
// Byte layouts (all lengths big-endian, strings UTF-8, no terminators):
//
//	hybrid v1:    BE32(len B0) ‖ B0 ‖ BE32(len A) ‖ A
//	hybrid v2:    "qorechain-pqc-hybrid-v2" ‖ BE64(len chainID) ‖ chainID ‖
//	              BE32(len B0) ‖ B0 ‖ BE32(len A) ‖ A
//	migration v1: "qorechain-key-migration:chain=%s:from=%d:to=%d:account=%s:height=%d"
//	migration v2: "qorechain-key-migration-v2" ‖ BE64(len chainID) ‖ chainID ‖
//	              BE64(len account) ‖ account ‖ BE32(from) ‖ BE32(to) ‖ BE64(height) ‖
//	              BE32(len oldPub) ‖ oldPub ‖ BE32(len newPub) ‖ newPub
//	bridge v1:    chain|eventType|operationID|txHash|amount|asset   (no chain id)
//	bridge v2:    "qorechain-bridge-attestation-v2" then, for each of
//	              [chainID, chain, eventType, operationID, txHash, amount, asset]:
//	              BE64(len f) ‖ f
//
// B0 is the TxBody serialised WITHOUT the PQC extension option and A is the
// AuthInfo bytes verbatim.
//
// This package depends only on the standard library so that every signing
// package (tx, unified, pqcdx) can import it without cycles.
package signbytes

import (
	"encoding/binary"
	"errors"
	"fmt"
	"strings"
)

// Domain tags and the upgrade that switches legacy networks to v2.
const (
	// HybridDomain prefixes the v2 hybrid tx sign-bytes.
	HybridDomain = "qorechain-pqc-hybrid-v2"
	// MigrationDomain prefixes the v2 PQC key-migration sign-bytes.
	MigrationDomain = "qorechain-key-migration-v2"
	// BridgeDomain prefixes the v2 bridge-attestation sign-bytes.
	BridgeDomain = "qorechain-bridge-attestation-v2"
	// V2Upgrade is the upgrade plan whose application switches a legacy network
	// from v1 to v2. Resolvers query /cosmos/upgrade/v1beta1/applied_plan/{V2Upgrade}.
	V2Upgrade = "v3.1.98"
)

// Version selects a sign-bytes form.
//
// V1 and V2 are resolved versions. Auto (or the zero value "") asks the SDK to
// decide: offline builders decide only when the chain id alone is enough (a
// non-legacy chain is always V2) and fail otherwise; a Resolver additionally
// asks the node whether V2Upgrade has been applied.
type Version string

// Sign-bytes versions.
const (
	// Auto resolves the version per network (the default).
	Auto Version = "auto"
	// V1 is the original form, verified by legacy networks before V2Upgrade.
	V1 Version = "v1"
	// V2 is the domain-tagged, chain-bound form.
	V2 Version = "v2"
)

// ErrUnresolvedVersion is returned when the version for a legacy network cannot
// be decided without asking the node (or the node could not be asked). Callers
// must pass a REST URL to a Resolver, or an explicit V1 / V2. The SDK never
// guesses.
var ErrUnresolvedVersion = errors.New("signbytes: cannot decide the hybrid sign-bytes version")

// legacyChains are the networks that ran before v2 existed and switch to it only
// when V2Upgrade is applied on them.
var legacyChains = map[string]bool{
	"qorechain-vladi": true,
	"qorechain-diana": true,
}

// LegacyChains returns the chain ids that start on v1 and switch to v2 only at
// V2Upgrade.
func LegacyChains() []string { return []string{"qorechain-vladi", "qorechain-diana"} }

// IsLegacyChain reports whether chainID is a network that verifies v1 until
// V2Upgrade is applied on it.
func IsLegacyChain(chainID string) bool { return legacyChains[chainID] }

// VersionFor mirrors the chain's own switch: V2 once V2Upgrade has been applied
// (v2AppliedHeight > 0) or for any chain that is not a legacy network; V1
// otherwise.
func VersionFor(chainID string, v2AppliedHeight int64) Version {
	if v2AppliedHeight > 0 || !IsLegacyChain(chainID) {
		return V2
	}
	return V1
}

// ParseVersion accepts "auto", "v1", "v2" (case-insensitive, surrounding space
// ignored) and "" (= Auto).
func ParseVersion(s string) (Version, error) {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "", "auto":
		return Auto, nil
	case "v1":
		return V1, nil
	case "v2":
		return V2, nil
	}
	return "", fmt.Errorf("signbytes: version must be auto, v1 or v2, got %q", s)
}

// IsAuto reports whether v asks for per-network resolution ("" or Auto).
func (v Version) IsAuto() bool { return v == "" || v == Auto }

// Validate checks that v is one of "", Auto, V1, V2.
func (v Version) Validate() error {
	switch v {
	case "", Auto, V1, V2:
		return nil
	}
	return fmt.Errorf("signbytes: version must be auto, v1 or v2, got %q", string(v))
}

// ResolveOffline decides the version without any network access. An explicit V1
// or V2 is returned as-is. For Auto (or ""), a non-legacy chain resolves to V2;
// a legacy chain cannot be decided offline and returns an error wrapping
// ErrUnresolvedVersion telling the caller to resolve against a node (Resolver)
// or pass V1 / V2 explicitly. It never defaults silently to V1.
func ResolveOffline(chainID string, requested Version) (Version, error) {
	if err := requested.Validate(); err != nil {
		return "", err
	}
	if !requested.IsAuto() {
		return requested, nil
	}
	if !IsLegacyChain(chainID) {
		return V2, nil
	}
	return "", fmt.Errorf("%w for chain %q: it verifies v1 or v2 depending on whether upgrade %s is applied; "+
		"resolve it with a signbytes.Resolver (REST URL) or pass signbytes.V1 / signbytes.V2 explicitly",
		ErrUnresolvedVersion, chainID, V2Upgrade)
}

// --- hybrid tx ---

// HybridV1 returns the v1 hybrid sign-bytes: BE32(len B0) ‖ B0 ‖ BE32(len A) ‖ A.
func HybridV1(b0, authInfo []byte) []byte {
	out := make([]byte, 0, 8+len(b0)+len(authInfo))
	out = binary.BigEndian.AppendUint32(out, uint32(len(b0)))
	out = append(out, b0...)
	out = binary.BigEndian.AppendUint32(out, uint32(len(authInfo)))
	out = append(out, authInfo...)
	return out
}

// HybridV2 returns the v2 hybrid sign-bytes:
// HybridDomain ‖ BE64(len chainID) ‖ chainID ‖ BE32(len B0) ‖ B0 ‖ BE32(len A) ‖ A.
func HybridV2(chainID string, b0, authInfo []byte) []byte {
	out := make([]byte, 0, len(HybridDomain)+8+len(chainID)+8+len(b0)+len(authInfo))
	out = append(out, HybridDomain...)
	out = binary.BigEndian.AppendUint64(out, uint64(len(chainID)))
	out = append(out, chainID...)
	out = binary.BigEndian.AppendUint32(out, uint32(len(b0)))
	out = append(out, b0...)
	out = binary.BigEndian.AppendUint32(out, uint32(len(authInfo)))
	out = append(out, authInfo...)
	return out
}

// Hybrid dispatches to HybridV1 / HybridV2. version must be resolved (V1 or V2);
// Auto is rejected so a caller cannot sign an undecided form.
func Hybrid(version Version, chainID string, b0, authInfo []byte) ([]byte, error) {
	switch version {
	case V1:
		return HybridV1(b0, authInfo), nil
	case V2:
		return HybridV2(chainID, b0, authInfo), nil
	}
	return nil, unresolvedArg(version)
}

// --- PQC key migration ---

// MigrationV1 returns the v1 (legacy, ASCII) key-migration sign-bytes:
// "qorechain-key-migration:chain=<chainID>:from=<from>:to=<to>:account=<account>:height=<height>".
func MigrationV1(chainID, account string, fromAlgo, toAlgo uint32, height int64) []byte {
	return []byte(fmt.Sprintf(
		"qorechain-key-migration:chain=%s:from=%d:to=%d:account=%s:height=%d",
		chainID, fromAlgo, toAlgo, account, height,
	))
}

// MigrationV2 returns the v2 key-migration sign-bytes, binding both public keys:
// MigrationDomain ‖ BE64(len chainID) ‖ chainID ‖ BE64(len account) ‖ account ‖
// BE32(from) ‖ BE32(to) ‖ BE64(height) ‖ BE32(len oldPub) ‖ oldPub ‖
// BE32(len newPub) ‖ newPub.
func MigrationV2(chainID, account string, fromAlgo, toAlgo uint32, height int64, oldPub, newPub []byte) []byte {
	out := make([]byte, 0, len(MigrationDomain)+8+len(chainID)+8+len(account)+4+4+8+4+len(oldPub)+4+len(newPub))
	out = append(out, MigrationDomain...)
	out = binary.BigEndian.AppendUint64(out, uint64(len(chainID)))
	out = append(out, chainID...)
	out = binary.BigEndian.AppendUint64(out, uint64(len(account)))
	out = append(out, account...)
	out = binary.BigEndian.AppendUint32(out, fromAlgo)
	out = binary.BigEndian.AppendUint32(out, toAlgo)
	out = binary.BigEndian.AppendUint64(out, uint64(height))
	out = binary.BigEndian.AppendUint32(out, uint32(len(oldPub)))
	out = append(out, oldPub...)
	out = binary.BigEndian.AppendUint32(out, uint32(len(newPub)))
	out = append(out, newPub...)
	return out
}

// Migration dispatches to MigrationV1 / MigrationV2 (oldPub/newPub are ignored
// by v1). version must be resolved (V1 or V2).
func Migration(version Version, chainID, account string, fromAlgo, toAlgo uint32, height int64, oldPub, newPub []byte) ([]byte, error) {
	switch version {
	case V1:
		return MigrationV1(chainID, account, fromAlgo, toAlgo, height), nil
	case V2:
		return MigrationV2(chainID, account, fromAlgo, toAlgo, height, oldPub, newPub), nil
	}
	return nil, unresolvedArg(version)
}

// --- bridge attestation ---

// BridgeAttestation carries the MsgBridgeAttestation fields that are signed.
// Amount is the decimal integer string exactly as the chain renders it
// (math.Int.String(), e.g. "1000000"; zero is "0").
type BridgeAttestation struct {
	Chain       string
	EventType   string
	OperationID string
	TxHash      string
	Amount      string
	Asset       string
}

// BridgeV1 returns the v1 (legacy) attestation sign-bytes: the six fields
// pipe-joined, with no chain id.
func BridgeV1(att BridgeAttestation) []byte {
	return []byte(fmt.Sprintf("%s|%s|%s|%s|%s|%s",
		att.Chain, att.EventType, att.OperationID, att.TxHash, att.Amount, att.Asset))
}

// BridgeV2 returns the v2 attestation sign-bytes: BridgeDomain followed by
// BE64(len f) ‖ f for f in [chainID, chain, eventType, operationID, txHash,
// amount, asset].
func BridgeV2(chainID string, att BridgeAttestation) []byte {
	fields := []string{chainID, att.Chain, att.EventType, att.OperationID, att.TxHash, att.Amount, att.Asset}
	n := len(BridgeDomain)
	for _, f := range fields {
		n += 8 + len(f)
	}
	out := make([]byte, 0, n)
	out = append(out, BridgeDomain...)
	for _, f := range fields {
		out = binary.BigEndian.AppendUint64(out, uint64(len(f)))
		out = append(out, f...)
	}
	return out
}

// Bridge dispatches to BridgeV1 / BridgeV2. version must be resolved (V1 or V2).
func Bridge(version Version, chainID string, att BridgeAttestation) ([]byte, error) {
	switch version {
	case V1:
		return BridgeV1(att), nil
	case V2:
		return BridgeV2(chainID, att), nil
	}
	return nil, unresolvedArg(version)
}

func unresolvedArg(v Version) error {
	if v.IsAuto() {
		return fmt.Errorf("%w: a builder needs a resolved version (v1 or v2), got %q; "+
			"resolve it first with ResolveOffline or a Resolver", ErrUnresolvedVersion, string(v))
	}
	return fmt.Errorf("signbytes: version must be v1 or v2, got %q", string(v))
}

// --- rejection detection ---

// HybridRejectionMessage is the chain's message for pqc codespace code 21.
const HybridRejectionMessage = "hybrid PQC signature verification failed"

// HybridRejectionCodespace / HybridRejectionCode identify the chain's refusal of
// a hybrid signature (for example one signed over the wrong sign-bytes form).
const (
	HybridRejectionCodespace = "pqc"
	HybridRejectionCode      = 21
)

// IsHybridRejection reports whether a broadcast result is the chain refusing the
// hybrid PQC signature: codespace "pqc" with code 21, or a log / message that
// contains HybridRejectionMessage. Code 21 in any other codespace (e.g. the root
// "sdk" codespace's "tx too large") is NOT this case.
func IsHybridRejection(codespace string, code uint32, log string) bool {
	if codespace == HybridRejectionCodespace && code == HybridRejectionCode {
		return true
	}
	return strings.Contains(log, HybridRejectionMessage)
}

File: packages/java/src/main/java/io/github/qorechain/tx/SignBytes.java
package io.github.qorechain.tx;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * Per-network post-quantum sign-bytes: the exact byte strings an ML-DSA-87 key
 * signs for a hybrid transaction, a PQC key migration, and a bridge attestation.
 *
 * <p>Each network verifies exactly ONE form at any height, never both:
 *
 * <ul>
 *   <li><b>v1</b> — the original forms. Verified by networks that have applied none
 *       of the {@link #SIGN_BYTES_V2_UPGRADES} plans (today: mainnet
 *       {@code qorechain-vladi}).
 *   <li><b>v2</b> — domain-tagged and chain-id-bound. Verified once one of those plans
 *       is applied (today: testnet {@code qorechain-diana}, which applied
 *       {@code v3.1.98} at height 5,746,000) and by every network born on that release
 *       or later.
 * </ul>
 *
 * <p>The switch ships under TWO plan names, {@code v3.2.0} (mainnet) and
 * {@code v3.1.98} (already applied on the testnet), both in
 * {@link #SIGN_BYTES_V2_UPGRADES}. A client must ask the network about every one of
 * them; asking a single name resolves v1 on the other network and every hybrid
 * transaction there is refused with {@code pqc} code 21.
 *
 * <p>Byte layouts (all lengths big-endian, strings UTF-8, no terminators):
 *
 * <pre>
 *   hybrid v1    = BE32(len B0) ‖ B0 ‖ BE32(len A) ‖ A
 *   hybrid v2    = "qorechain-pqc-hybrid-v2" ‖ BE64(len chainId) ‖ chainId ‖
 *                  BE32(len B0) ‖ B0 ‖ BE32(len A) ‖ A
 *   migration v1 = "qorechain-key-migration:chain=%s:from=%d:to=%d:account=%s:height=%d"
 *   migration v2 = "qorechain-key-migration-v2" ‖ BE64(len chainId) ‖ chainId ‖
 *                  BE64(len account) ‖ account ‖ BE32(from) ‖ BE32(to) ‖ BE64(height) ‖
 *                  BE32(len oldPub) ‖ oldPub ‖ BE32(len newPub) ‖ newPub
 *   bridge v1    = chain|eventType|operationId|txHash|amount|asset   (no chain id)
 *   bridge v2    = "qorechain-bridge-attestation-v2" ‖ for f in
 *                  [chainId, chain, eventType, operationId, txHash, amount, asset]:
 *                  BE64(len f) ‖ f
 * </pre>
 *
 * <p>Which form to sign is decided by {@link #versionFor} (pure) or, with a network
 * lookup, by {@link SignBytesResolver}. Everything in this class is pure.
 */
public final class SignBytes {

    private SignBytes() {}

    /** Domain tag of the v2 hybrid transaction sign-bytes. */
    public static final String HYBRID_DOMAIN = "qorechain-pqc-hybrid-v2";

    /** Domain tag of the v2 key-migration sign-bytes. */
    public static final String MIGRATION_DOMAIN = "qorechain-key-migration-v2";

    /** Domain tag of the v2 bridge-attestation sign-bytes. */
    public static final String BRIDGE_DOMAIN = "qorechain-bridge-attestation-v2";

    /**
     * The PRIMARY name of the coordinated upgrade whose handler switches a pre-existing
     * network to v2 — the name mainnet ({@code qorechain-vladi}) applies.
     *
     * <p>It is <b>not</b> the only one. A client must ask the network about every name
     * in {@link #SIGN_BYTES_V2_UPGRADES}, never this one alone: the testnet took the
     * same batch under the earlier name {@code v3.1.98} and keeps that record forever,
     * so a single-name lookup answers "not applied" on one of the two networks and
     * resolves v1 there — the form it refuses with {@code pqc} code 21.
     */
    public static final String SIGN_BYTES_V2_UPGRADE = "v3.2.0";

    /**
     * Every upgrade plan name whose application switches a legacy network from v1 to
     * v2, most likely first.
     *
     * <p>The chain registers ONE handler under all of these names (qorechain-core
     * {@code x/pqc/types.SignBytesV2Upgrades}): mainnet applies {@code v3.2.0}, while
     * the testnet already applied {@code v3.1.98} and keeps answering under it. A
     * resolver queries {@code /cosmos/upgrade/v1beta1/applied_plan/{name}} for each
     * name in this order and takes v2 as soon as one answers a height greater than
     * zero.
     */
    public static final List<String> SIGN_BYTES_V2_UPGRADES = List.of("v3.2.0", "v3.1.98");

    /**
     * Networks that were running before v2 existed; they switch to v2 only once one of
     * {@link #SIGN_BYTES_V2_UPGRADES} is applied. Any other chain id is v2 from its
     * first block.
     */
    public static final Set<String> LEGACY_CHAINS = Set.of("qorechain-vladi", "qorechain-diana");

    /** The pqc codespace error code for a refused hybrid signature. */
    public static final int PQC_HYBRID_VERIFY_FAILED_CODE = 21;

    /** The chain's log text for a refused hybrid signature. */
    public static final String PQC_HYBRID_VERIFY_FAILED_LOG =
            "hybrid PQC signature verification failed";

    /** A resolved sign-bytes form. */
    public enum Version {
        V1(1),
        V2(2);

        private final int number;

        Version(int number) {
            this.number = number;
        }

        /** 1 or 2. */
        public int number() {
            return number;
        }

        /** {@code "v1"} or {@code "v2"}. */
        public String wireName() {
            return "v" + number;
        }
    }

    /**
     * A caller's choice: {@link #AUTO} asks the network (via
     * {@link SignBytesResolver}); {@link #V1}/{@link #V2} force a form with no network
     * call.
     */
    public enum Mode {
        AUTO,
        V1,
        V2;

        /** Parse {@code "auto"}, {@code "v1"} or {@code "v2"} (case-insensitive; null/empty = AUTO). */
        public static Mode parse(String s) {
            if (s == null || s.isEmpty()) {
                return AUTO;
            }
            switch (s.toLowerCase(Locale.ROOT)) {
                case "auto":
                    return AUTO;
                case "v1":
                    return V1;
                case "v2":
                    return V2;
                default:
                    throw new IllegalArgumentException(
                            "sign-bytes mode must be auto, v1 or v2, got \"" + s + "\"");
            }
        }

        /** The explicit version for V1/V2, or {@code null} for AUTO. */
        public Version explicitVersion() {
            switch (this) {
                case V1:
                    return Version.V1;
                case V2:
                    return Version.V2;
                default:
                    return null;
            }
        }
    }

    // ---- version selection ----

    /** Whether {@code chainId} is one of the pre-v2 networks in {@link #LEGACY_CHAINS}. */
    public static boolean isLegacyChain(String chainId) {
        return chainId != null && LEGACY_CHAINS.contains(chainId);
    }

    /**
     * The form a client must sign for {@code chainId}, given the height at which one
     * of {@link #SIGN_BYTES_V2_UPGRADES} was applied on that chain — the greatest
     * height any of the names answered, 0 when none has been applied. Mirrors the
     * chain's {@code SignBytesVersionFor} exactly: v2 when the height is positive OR
     * the chain is not a legacy network; v1 otherwise.
     */
    public static Version versionFor(String chainId, long v2AppliedHeight) {
        if (v2AppliedHeight > 0 || !isLegacyChain(chainId)) {
            return Version.V2;
        }
        return Version.V1;
    }

    /**
     * The version a pure (synchronous, network-free) builder must use.
     *
     * <p>An explicit {@code version} is returned as-is. When it is {@code null}, a
     * non-legacy chain is unambiguously v2; a legacy chain ({@code qorechain-vladi},
     * {@code qorechain-diana}) cannot be decided without asking the network, so this
     * throws rather than guessing.
     *
     * @throws IllegalStateException for a legacy chain with no explicit version.
     */
    public static Version requireVersion(String chainId, Version version) {
        if (version != null) {
            return version;
        }
        if (!isLegacyChain(chainId)) {
            return Version.V2;
        }
        throw new IllegalStateException(
                "hybrid sign-bytes version is required for chain \""
                        + chainId
                        + "\": it signs v1 until one of the "
                        + upgradeNames()
                        + " upgrades is applied and v2 after. Pass an explicit"
                        + " signBytesVersion (V1/V2), or resolve it with SignBytesResolver"
                        + " against the network's REST endpoint.");
    }

    /** {@link #SIGN_BYTES_V2_UPGRADES} rendered for an error message ("v3.2.0 / v3.1.98"). */
    static String upgradeNames() {
        return String.join(" / ", SIGN_BYTES_V2_UPGRADES);
    }

    // ---- hybrid ----

    /** Hybrid v1: {@code BE32(len b0) ‖ b0 ‖ BE32(len authInfo) ‖ authInfo}. */
    public static byte[] hybridV1(byte[] b0, byte[] authInfo) {
        ByteArrayOutputStream out = new ByteArrayOutputStream(8 + b0.length + authInfo.length);
        out.writeBytes(be32(b0.length));
        out.writeBytes(b0);
        out.writeBytes(be32(authInfo.length));
        out.writeBytes(authInfo);
        return out.toByteArray();
    }

    /**
     * Hybrid v2: {@code "qorechain-pqc-hybrid-v2" ‖ BE64(len chainId) ‖ chainId ‖
     * BE32(len b0) ‖ b0 ‖ BE32(len authInfo) ‖ authInfo}.
     */
    public static byte[] hybridV2(String chainId, byte[] b0, byte[] authInfo) {
        requireChainId(chainId);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.writeBytes(utf8(HYBRID_DOMAIN));
        writeBe64Prefixed(out, utf8(chainId));
        out.writeBytes(be32(b0.length));
        out.writeBytes(b0);
        out.writeBytes(be32(authInfo.length));
        out.writeBytes(authInfo);
        return out.toByteArray();
    }

    /** Hybrid sign-bytes in the given form ({@code chainId} is ignored by v1). */
    public static byte[] hybrid(Version version, String chainId, byte[] b0, byte[] authInfo) {
        if (version == null) {
            throw new IllegalArgumentException("sign-bytes version is required");
        }
        return version == Version.V2 ? hybridV2(chainId, b0, authInfo) : hybridV1(b0, authInfo);
    }

    // ---- key migration ----

    /**
     * Migration v1 (legacy ASCII):
     * {@code qorechain-key-migration:chain=%s:from=%d:to=%d:account=%s:height=%d}.
     * Algorithm ids are formatted unsigned (they are {@code uint32} on chain).
     */
    public static byte[] migrationV1(
            String chainId, String account, int fromAlgorithmId, int toAlgorithmId, long height) {
        return utf8(
                "qorechain-key-migration:chain="
                        + chainId
                        + ":from="
                        + Integer.toUnsignedString(fromAlgorithmId)
                        + ":to="
                        + Integer.toUnsignedString(toAlgorithmId)
                        + ":account="
                        + account
                        + ":height="
                        + height);
    }

    /** Migration v2 (see the class header for the layout). */
    public static byte[] migrationV2(
            String chainId,
            String account,
            int fromAlgorithmId,
            int toAlgorithmId,
            long height,
            byte[] oldPublicKey,
            byte[] newPublicKey) {
        requireChainId(chainId);
        byte[] oldPub = oldPublicKey == null ? new byte[0] : oldPublicKey;
        byte[] newPub = newPublicKey == null ? new byte[0] : newPublicKey;
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.writeBytes(utf8(MIGRATION_DOMAIN));
        writeBe64Prefixed(out, utf8(chainId));
        writeBe64Prefixed(out, utf8(account));
        out.writeBytes(be32(fromAlgorithmId));
        out.writeBytes(be32(toAlgorithmId));
        out.writeBytes(be64(height));
        out.writeBytes(be32(oldPub.length));
        out.writeBytes(oldPub);
        out.writeBytes(be32(newPub.length));
        out.writeBytes(newPub);
        return out.toByteArray();
    }

    /** Migration sign-bytes in the given form (v1 ignores the public keys). */
    public static byte[] migration(
            Version version,
            String chainId,
            String account,
            int fromAlgorithmId,
            int toAlgorithmId,
            long height,
            byte[] oldPublicKey,
            byte[] newPublicKey) {
        if (version == null) {
            throw new IllegalArgumentException("sign-bytes version is required");
        }
        return version == Version.V2
                ? migrationV2(
                        chainId,
                        account,
                        fromAlgorithmId,
                        toAlgorithmId,
                        height,
                        oldPublicKey,
                        newPublicKey)
                : migrationV1(chainId, account, fromAlgorithmId, toAlgorithmId, height);
    }

    // ---- bridge attestation ----

    /**
     * Bridge attestation v1 (legacy ASCII, no chain id):
     * {@code chain|eventType|operationId|txHash|amount|asset}. {@code amount} is the
     * decimal integer string.
     */
    public static byte[] bridgeV1(
            String chain,
            String eventType,
            String operationId,
            String txHash,
            String amount,
            String asset) {
        return utf8(
                String.join("|", chain, eventType, operationId, txHash, amount, asset));
    }

    /** Bridge attestation v2 (see the class header for the layout). */
    public static byte[] bridgeV2(
            String chainId,
            String chain,
            String eventType,
            String operationId,
            String txHash,
            String amount,
            String asset) {
        requireChainId(chainId);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.writeBytes(utf8(BRIDGE_DOMAIN));
        for (String f : new String[] {chainId, chain, eventType, operationId, txHash, amount, asset}) {
            writeBe64Prefixed(out, utf8(f));
        }
        return out.toByteArray();
    }

    /** Bridge attestation sign-bytes in the given form (v1 ignores {@code chainId}). */
    public static byte[] bridge(
            Version version,
            String chainId,
            String chain,
            String eventType,
            String operationId,
            String txHash,
            String amount,
            String asset) {
        if (version == null) {
            throw new IllegalArgumentException("sign-bytes version is required");
        }
        return version == Version.V2
                ? bridgeV2(chainId, chain, eventType, operationId, txHash, amount, asset)
                : bridgeV1(chain, eventType, operationId, txHash, amount, asset);
    }

    // ---- rejection detection ----

    /**
     * Whether a broadcast refusal is the chain rejecting the hybrid PQC signature —
     * the symptom of signing the wrong sign-bytes form. True when
     * {@code codespace == "pqc" && code == 21}, or when the log/message contains
     * {@value #PQC_HYBRID_VERIFY_FAILED_LOG}. Code 21 in any other codespace is NOT
     * this case.
     */
    public static boolean isHybridVerifyRejection(int code, String codespace, String log) {
        if ("pqc".equals(codespace) && code == PQC_HYBRID_VERIFY_FAILED_CODE) {
            return true;
        }
        return log != null && log.contains(PQC_HYBRID_VERIFY_FAILED_LOG);
    }

    /** {@link #isHybridVerifyRejection(int, String, String)} over a thrown broadcast error. */
    public static boolean isHybridVerifyRejection(Throwable error) {
        if (error instanceof TxError.QoreTxException) {
            TxError.QoreTxException e = (TxError.QoreTxException) error;
            if (isHybridVerifyRejection(e.code, e.codespace, e.rawLog)) {
                return true;
            }
        }
        return error != null
                && error.getMessage() != null
                && error.getMessage().contains(PQC_HYBRID_VERIFY_FAILED_LOG);
    }

    // ---- encoding helpers ----

    /** Big-endian 4-byte encoding of {@code n} (as an unsigned 32-bit value). */
    public static byte[] be32(int n) {
        return new byte[] {
            (byte) (n >>> 24), (byte) (n >>> 16), (byte) (n >>> 8), (byte) n
        };
    }

    /** Big-endian 8-byte encoding of {@code n}. */
    public static byte[] be64(long n) {
        byte[] out = new byte[8];
        for (int i = 7; i >= 0; i--) {
            out[i] = (byte) n;
            n >>>= 8;
        }
        return out;
    }

    private static void writeBe64Prefixed(ByteArrayOutputStream out, byte[] field) {
        out.writeBytes(be64(field.length));
        out.writeBytes(field);
    }

    private static byte[] utf8(String s) {
        return (s == null ? "" : s).getBytes(StandardCharsets.UTF_8);
    }

    private static void requireChainId(String chainId) {
        if (chainId == null) {
            throw new IllegalArgumentException("chainId is required for v2 sign-bytes");
        }
    }
}
